File: src/core/cmd/events/keyboard.rs
use serde::{Deserialize, Serialize};

use super::common::{ElementState, ModifiersState};

/// Key location on keyboard
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KeyLocation {
    Standard = 0,
    Left = 1,
    Right = 2,
    Numpad = 3,
}

/// Physical key code (scancode-like, layout independent)
#[repr(u16)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KeyCode {
    // Writing System Keys
    Backquote,
    Backslash,
    BracketLeft,
    BracketRight,
    Comma,
    Digit0,
    Digit1,
    Digit2,
    Digit3,
    Digit4,
    Digit5,
    Digit6,
    Digit7,
    Digit8,
    Digit9,
    Equal,
    IntlBackslash,
    IntlRo,
    IntlYen,
    KeyA,
    KeyB,
    KeyC,
    KeyD,
    KeyE,
    KeyF,
    KeyG,
    KeyH,
    KeyI,
    KeyJ,
    KeyK,
    KeyL,
    KeyM,
    KeyN,
    KeyO,
    KeyP,
    KeyQ,
    KeyR,
    KeyS,
    KeyT,
    KeyU,
    KeyV,
    KeyW,
    KeyX,
    KeyY,
    KeyZ,
    Minus,
    Period,
    Quote,
    Semicolon,
    Slash,

    // Functional Keys
    AltLeft,
    AltRight,
    Backspace,
    CapsLock,
    ContextMenu,
    ControlLeft,
    ControlRight,
    Enter,
    SuperLeft,
    SuperRight,
    ShiftLeft,
    ShiftRight,
    Space,
    Tab,

    // Control Keys
    Delete,
    End,
    Help,
    Home,
    Insert,
    PageDown,
    PageUp,

    // Arrow Keys
    ArrowDown,
    ArrowLeft,
    ArrowRight,
    ArrowUp,

    // Numpad Keys
    NumLock,
    Numpad0,
    Numpad1,
    Numpad2,
    Numpad3,
    Numpad4,
    Numpad5,
    Numpad6,
    Numpad7,
    Numpad8,
    Numpad9,
    NumpadAdd,
    NumpadBackspace,
    NumpadClear,
    NumpadClearEntry,
    NumpadComma,
    NumpadDecimal,
    NumpadDivide,
    NumpadEnter,
    NumpadEqual,
    NumpadHash,
    NumpadMemoryAdd,
    NumpadMemoryClear,
    NumpadMemoryRecall,
    NumpadMemoryStore,
    NumpadMemorySubtract,
    NumpadMultiply,
    NumpadParenLeft,
    NumpadParenRight,
    NumpadStar,
    NumpadSubtract,

    // Function Keys
    Escape,
    F1,
    F2,
    F3,
    F4,
    F5,
    F6,
    F7,
    F8,
    F9,
    F10,
    F11,
    F12,
    F13,
    F14,
    F15,
    F16,
    F17,
    F18,
    F19,
    F20,
    F21,
    F22,
    F23,
    F24,

    // Lock Keys
    ScrollLock,

    // Media Keys
    AudioVolumeDown,
    AudioVolumeMute,
    AudioVolumeUp,
    MediaPlayPause,
    MediaStop,
    MediaTrackNext,
    MediaTrackPrevious,

    // Browser Keys
    BrowserBack,
    BrowserFavorites,
    BrowserForward,
    BrowserHome,
    BrowserRefresh,
    BrowserSearch,
    BrowserStop,

    // System Keys
    PrintScreen,
    Pause,

    // Unknown/Unidentified key
    Unidentified,
}

/// Keyboard input event
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data", rename_all = "kebab-case")]
pub enum KeyboardEvent {
    /// Key was pressed or released
    OnInput {
        window_id: u32,
        key_code: KeyCode,
        state: ElementState,
        location: KeyLocation,
        repeat: bool,
        text: Option<String>,
        modifiers: ModifiersState,
    },

    /// Modifiers changed
    OnModifiersChange {
        window_id: u32,
        modifiers: ModifiersState,
    },

    /// IME composition started
    OnImeEnable { window_id: u32 },

    /// IME composition in progress
    OnImePreedit {
        window_id: u32,
        text: String,
        cursor_range: Option<(usize, usize)>,
    },

    /// IME composition committed
    OnImeCommit { window_id: u32, text: String },

    /// IME disabled
    OnImeDisable { window_id: u32 },
}

File: binding/src/events/keyboard.ts
import type { ElementState, ModifiersState } from './common';

// MARK: Keyboard Events

/** Key location on keyboard */
export type KeyLocation = 'standard' | 'left' | 'right' | 'numpad';

/** Physical key code (scancode-like, layout independent) */
export type KeyCode =
  // Writing System Keys
  | 'backquote'
  | 'backslash'
  | 'bracket-left'
  | 'bracket-right'
  | 'comma'
  | 'digit0'
  | 'digit1'
  | 'digit2'
  | 'digit3'
  | 'digit4'
  | 'digit5'
  | 'digit6'
  | 'digit7'
  | 'digit8'
  | 'digit9'
  | 'equal'
  | 'intl-backslash'
  | 'intl-ro'
  | 'intl-yen'
  | 'key-a'
  | 'key-b'
  | 'key-c'
  | 'key-d'
  | 'key-e'
  | 'key-f'
  | 'key-g'
  | 'key-h'
  | 'key-i'
  | 'key-j'
  | 'key-k'
  | 'key-l'
  | 'key-m'
  | 'key-n'
  | 'key-o'
  | 'key-p'
  | 'key-q'
  | 'key-r'
  | 'key-s'
  | 'key-t'
  | 'key-u'
  | 'key-v'
  | 'key-w'
  | 'key-x'
  | 'key-y'
  | 'key-z'
  | 'minus'
  | 'period'
  | 'quote'
  | 'semicolon'
  | 'slash'
  // Functional Keys
  | 'alt-left'
  | 'alt-right'
  | 'backspace'
  | 'caps-lock'
  | 'context-menu'
  | 'control-left'
  | 'control-right'
  | 'enter'
  | 'super-left'
  | 'super-right'
  | 'shift-left'
  | 'shift-right'
  | 'space'
  | 'tab'
  // Control Keys
  | 'delete'
  | 'end'
  | 'help'
  | 'home'
  | 'insert'
  | 'page-down'
  | 'page-up'
  // Arrow Keys
  | 'arrow-down'
  | 'arrow-left'
  | 'arrow-right'
  | 'arrow-up'
  // Numpad Keys
  | 'num-lock'
  | 'numpad0'
  | 'numpad1'
  | 'numpad2'
  | 'numpad3'
  | 'numpad4'
  | 'numpad5'
  | 'numpad6'
  | 'numpad7'
  | 'numpad8'
  | 'numpad9'
  | 'numpad-add'
  | 'numpad-backspace'
  | 'numpad-clear'
  | 'numpad-clear-entry'
  | 'numpad-comma'
  | 'numpad-decimal'
  | 'numpad-divide'
  | 'numpad-enter'
  | 'numpad-equal'
  | 'numpad-hash'
  | 'numpad-memory-add'
  | 'numpad-memory-clear'
  | 'numpad-memory-recall'
  | 'numpad-memory-store'
  | 'numpad-memory-subtract'
  | 'numpad-multiply'
  | 'numpad-paren-left'
  | 'numpad-paren-right'
  | 'numpad-star'
  | 'numpad-subtract'
  // Function Keys
  | 'escape'
  | 'f1'
  | 'f2'
  | 'f3'
  | 'f4'
  | 'f5'
  | 'f6'
  | 'f7'
  | 'f8'
  | 'f9'
  | 'f10'
  | 'f11'
  | 'f12'
  | 'f13'
  | 'f14'
  | 'f15'
  | 'f16'
  | 'f17'
  | 'f18'
  | 'f19'
  | 'f20'
  | 'f21'
  | 'f22'
  | 'f23'
  | 'f24'
  // Lock Keys
  | 'scroll-lock'
  // Media Keys
  | 'audio-volume-down'
  | 'audio-volume-mute'
  | 'audio-volume-up'
  | 'media-play-pause'
  | 'media-stop'
  | 'media-track-next'
  | 'media-track-previous'
  // Browser Keys
  | 'browser-back'
  | 'browser-favorites'
  | 'browser-forward'
  | 'browser-home'
  | 'browser-refresh'
  | 'browser-search'
  | 'browser-stop'
  // System Keys
  | 'print-screen'
  | 'pause'
  // Unknown
  | 'unidentified';

/**
 * Event fired when a keyboard key is pressed or released.
 *
 * @example
 * ```typescript
 * const event: KeyboardInputEvent = {
 *   event: 'on-input',
 *   data: {
 *     windowId: 1,
 *     keyCode: 'key-w',
 *     state: 'pressed',
 *     location: 'standard',
 *     repeat: false,
 *     text: 'w',
 *     modifiers: { shift: false, ctrl: false, alt: false, meta: false }
 *   }
 * };
 * ```
 */
export interface KeyboardInputEvent {
  event: 'on-input';
  data: {
    windowId: number;
    keyCode: KeyCode;
    state: ElementState;
    location: KeyLocation;
    repeat: boolean;
    text: string | null;
    modifiers: ModifiersState;
  };
}

export interface KeyboardModifiersChangedEvent {
  event: 'on-modifiers-change';
  data: {
    windowId: number;
    modifiers: ModifiersState;
  };
}

export interface KeyboardImeEnabledEvent {
  event: 'on-ime-enable';
  data: { windowId: number };
}

export interface KeyboardImePreeditEvent {
  event: 'on-ime-preedit';
  data: {
    windowId: number;
    text: string;
    cursorRange: [number, number] | null;
  };
}

export interface KeyboardImeCommitEvent {
  event: 'on-ime-commit';
  data: {
    windowId: number;
    text: string;
  };
}

export interface KeyboardImeDisabledEvent {
  event: 'on-ime-disable';
  data: { windowId: number };
}

export type KeyboardEvent =
  | KeyboardInputEvent
  | KeyboardModifiersChangedEvent
  | KeyboardImeEnabledEvent
  | KeyboardImePreeditEvent
  | KeyboardImeCommitEvent
  | KeyboardImeDisabledEvent;
